datalad/crawler/pipelines/balsa.py

```diff
@@ -0,0 +1,219 @@
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+
+"""A pipeline for crawling BALSA datasets"""
+
+import os
+from shutil import rmtree
+from os import curdir, makedirs, rmdir
+from os.path import lexists, join as opj, abspath, exists, normpath
+from collections import OrderedDict
+
+# Import necessary nodes
+from ..nodes.crawl_url import crawl_url
+from ..nodes.matches import xpath_match, a_href_match
+from ..nodes.misc import assign, skip_if, find_files, continue_if
+from ..nodes.misc import debug
+from ..nodes.misc import sub, switch
+from ..nodes.annex import Annexificator
+from ...consts import ARCHIVES_SPECIAL_REMOTE, DATALAD_SPECIAL_REMOTE
+from datalad.utils import find_files as f_f
+from datalad.support.annexrepo import *
+from datalad.support.gitrepo import *
+
+from logging import getLogger
+lgr = getLogger("datalad.crawler.pipelines.balsa")
+
+TOPURL = "https://balsa.wustl.edu/"
+
+
+def superdataset_pipeline(url=TOPURL):
+    """
+    Parameters
+    ----------
+    url: str
+       URL point to all datasets, hence the URL at the top
+
+    """
+    url = opj(url, 'study/')
+    hardcodeNames = {'RVVG': '2016-Glasser_et_al-Multimodal_parcellation',
+                     'W336': '2016-Donahue_et_al-Connection_strength_and_distance',
+                     'WG33': '2016-VanEssen_et_al-BALSA'
+                     }
+
+    annex = Annexificator()
+    lgr.info("Creating a BALSA collection pipeline")
+    return [
+        crawl_url(url),
+        a_href_match('.*/study/show/(?P<dataset_id>.*)'),
+        # skip the empty dataset used by BALSA for testing
+        skip_if({'dataset_id': 'Jvw1'}, re=True),
+        crawl_url(),
+        xpath_match('substring(//*/h3/text(), 8, string-length(//*/h3/text()))', output='dataset'),
+        # assign({'dataset_name': '%(dataset)s'}, interpolate=True),
+        switch('dataset_id', {idkey: assign({'dataset_name': name}) for idkey, name in hardcodeNames.items()}),
+        annex.initiate_dataset(
+            template="balsa",
+            data_fields=['dataset_id'],
+            existing='skip'
+        )
+    ]
+
+
+def extract_meta(data):
+    content = [x['match'] for x in xpath_match('//*[@class="attributeLabel"]/..')(data)]
+    content = [(re.sub('</li>', ', ', x)) for x in content]
+    content = [(re.sub('<[^<]+?>|[\t|\n|\r]', '', (str(x.encode('ascii', 'ignore'))))).strip() for x in content]
+    [content.remove(x) for x in content if x.find('SCENES:' or 'OWNERS:') >= 0]
+
+    json_dict = OrderedDict(map(str, x.split(':', 1)) for x in content)
+
+    if not exists(".datalad/meta"):
+        makedirs(".datalad/meta")
+
+    opj(".datalad/meta/balsa.json".split('/'))
```

Comment:
this line has no effect ;)  I guess it was just me demoing how you could take POSIX-compliant path (with / in it) and feed it back into `opj`.   But now we have got a little helper `_path_`. If you haven't merged the most recent origin/master, merge it, and then use it here (import from datalad.utils). e.g.
```with open(_path_(".datalad/meta/balsa.json"), "w") as fi:```
Eventually we will be compatible with Windows!  ;) so let's not keep loose points